Source/Scene/modelMaterialsCommon.js

```diff
@@ -0,0 +1,792 @@
+                ],
+                depthMask: false,
+                functions: {
+                    "blendEquationSeparate": [
+                        WebGLConstants.FUNC_ADD,
+                        WebGLConstants.FUNC_ADD
+                    ],
+                    "blendFuncSeparate": [
+                        WebGLConstants.ONE,
+                        WebGLConstants.ONE_MINUS_SRC_ALPHA,
+                        WebGLConstants.ONE,
+                        WebGLConstants.ONE_MINUS_SRC_ALPHA
+                    ]
+                }
+            };
+        }
+        else {
+            techniqueStates = {
+                enable: [
+                    WebGLConstants.CULL_FACE,
+                    WebGLConstants.DEPTH_TEST
+                ]
+            };
+        }
+        techniques[techniqueId] = {
+            attributes: techniqueAttributes,
+            parameters: techniqueParameters,
+            program: programId,
+            states: techniqueStates,
+            uniforms: techniqueUniforms
+        };
+
+        // Add shaders
+        shaders[vertexShaderId] = {
+            type: WebGLConstants.VERTEX_SHADER,
+            uri: '',
+            extras: {
+                source: vertexShader
+            }
+        };
+        shaders[fragmentShaderId] = {
+            type: WebGLConstants.FRAGMENT_SHADER,
+            uri: '',
+            extras: {
+                source: fragmentShader
+            }
+        };
+
+        // Add program
+        var programAttributes = [];
+        for (i=0;i<attributesCount;++i) {
+            programAttributes.push('a_' + attributes[i].toLowerCase());
+        }
+
+        programs[programId] = {
+            attributes: programAttributes,
+            fragmentShader: fragmentShaderId,
+            vertexShader: vertexShaderId
+        };
+
+        return techniqueId;
+    }
+
+    function getKey(khrMaterialsCommon) {
+        var key = '';
+        key += 'technique:' + khrMaterialsCommon.technique + ';';
+
+        var values = khrMaterialsCommon.values;
+        var keys = Object.keys(values).sort();
+        var keysCount = keys.length;
+        for (var i=0;i<keysCount;++i) {
+            var name = keys[i];
+            if (values.hasOwnProperty(name)) {
+                var value = values[name];
+                key += name + ':';
+                var type = typeof value;
+                switch (type) {
+                    case 'string':
+                        key += 'texture';
+                        break;
+                    case 'number':
+                        key += 'float';
+                        break;
+                    default:
+                        if (Array.isArray(value)) {
+                            key += 'vec' + value.length.toString();
+                        }
+                        break;
+                }
+                key += ';';
+            }
+        }
+    }
+
+    function getPrimitiveInfo(materialId, gltf) {
```

Comment:
This is awkward (and probably slow).  Would it be useful to have attribute information as part of the material extension?  For example, just like what a stock glTF technique has: https://github.com/KhronosGroup/glTF/blob/spec-1.0/specification/schema/examples/techniques.json